File: 257.binary-tree-paths.py
#
# @lc app=leetcode id=257 lang=python3
#
# [257] Binary Tree Paths
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:

        res = []
        self.dfs(root, res, "")
        return res
    
    
    def dfs(self, root, res, path):
        if not root:
            return
        if not root.left and not root.right:
            path += str(root.val)
            res.append(path)
            return
        
        path += "{}->".format(root.val)
        self.dfs(root.left, res, path)
        self.dfs(root.right, res, path)



        

        
# @lc code=end

File: 257.binary-tree-paths.java
/*
 * @lc app=leetcode id=257 lang=java
 *
 * [257] Binary Tree Paths
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> res = new ArrayList<>();
        dfs(root, res, "");
        return res;
    }

    public void dfs(TreeNode root, List<String> res, String path) {
        if (root == null) {
            return;
        }

        if (root.left == null && root.right == null) {
            path += root.val;
            res.add(path);
            return;
        }


        path += root.val + "->";
        dfs(root.left, res, path);
        dfs(root.right, res, path);
    }
}
// @lc code=end

